socorro/external/es/crashstorage.py

```diff
@@ -124,18 +183,62 @@ def get_index_for_crash(self, crash_date):
 
         return index
 
-    def save_raw_and_processed(self, raw_crash, dumps, processed_crash,
-                               crash_id):
-        """This is the only write mechanism that is actually employed in normal
-        usage.
+    def save_raw_and_processed(self, raw_crash, dumps, processed_crash, crash_id):
+        """Save raw and processed crash data to Elasticsearch
+
+        Note: This is the only write mechanism that is actually employed in
+        normal usage.
+
         """
 
+        # Massage the crash such that the date_processed field is formatted
+        # in the fashion of our established mapping.
+        self.reconstitute_datetimes(processed_crash)
+
+        # Remove bad keys from the raw crash--these keys are essentially
+        # user-provided and can contain junk data
+        remove_bad_keys(raw_crash)
+
+        # Truncate values that are too long
+        truncate_keyword_field_values(FIELDS, raw_crash)
+        truncate_keyword_field_values(FIELDS, processed_crash)
+
+        # Capture crash data size metrics--do this only after we've cleaned up
+        # the crash data
+        self._capture_crash_metrics(raw_crash, processed_crash)
+
         crash_document = {
             'crash_id': crash_id,
             'processed_crash': processed_crash,
             'raw_crash': raw_crash
         }
 
+        self.transaction(
+            self._submit_crash_to_elasticsearch,
+            crash_document=crash_document
+        )
+
+    @staticmethod
+    def reconstitute_datetimes(processed_crash):
+        # FIXME(willkg): These should be specified in super_search_fields.py
+        # and not hard-coded
+        datetime_fields = [
+            'submitted_timestamp',
+            'date_processed',
+            'client_crash_date',
+            'started_datetime',
+            'startedDateTime',
+            'completed_datetime',
+            'completeddatetime',
+        ]
+        for a_key in datetime_fields:
+            if a_key not in processed_crash:
+                continue
+
+            processed_crash[a_key] = string_to_datetime(processed_crash[a_key])
+
+    def _capture_crash_metrics(self, raw_crash, processed_crash):
```

Comment:
Do we need an underscore here? I can't really see a pattern for their use in this class.

Reply:
I put an `_` there because other things in the class that are internal to the class and aren't used in the tests have one. I don't think it matters a whole lot here and it's probably not a helpful signal. I'll nix it.